### PR TITLE
Issue Fixes for Data Validation

### PR DESCRIPTION
**Pull request by KinanBab**
This includes fixes for #42 #43 #48 #51 

Comments about #48 :
* I made a tiny modification to the code that parses the spreadsheet. I made the workend callback get a flag parameter to indicate whether errors occurred or not.
* The code for reading spreadsheets perform some validation to check if cells are not numbers using (isNaN function). It does not rely on the validators of the table. The behaviour of isNaN is a sub-set of the behaviour of the table validators (the table validators are stricter). 
* If there are errors thrown at the parsing stage (whether because of format issues or because of the isNaN validation), then the tables view is not expanded.
* If there are no errors thrown at the parsing stage, then the tables view is expanded, even if there are validation errors from the table validators.
* If the desired behaviour is to not expand the view when there are validation errors from table validators as well, a quick fix would be to validate tables inside workend before expanding.
* On the long run, I think it would be better for the spread sheet parsing code to rely on the table validators, this way the spread sheet parsing code can work with any generic table definition.

### CODE REVIEW

File: client/index.html [hunk 1]
```diff
@@ -493,6 +418,7 @@ <h4 id="totals-hot-name"></h4>
             return o;
           }(cols));
           calculateSize();
+          alert("HELLO");
```
Comment (frederickjansen) [inline client/index.html:434]: Remove alert

File: client/script/client.js [hunk 2]
```diff
@@ -17,11 +19,41 @@ var client = (function () {
...
+      return false;
+    }
+  };
+
+  function verifyKeysAndFetchDescription(callback) {
     var session = $("#session").val().trim().toLowerCase();
     var participationCode = $("#participation-code").val().trim().toLowerCase();
 
-    if (session === "" || participationCode === "") return;
+    if (session == "" || participationCode == "") {
```
Comment (frederickjansen) [inline client/script/client.js:54]: Not sure why you're reverting to type coercion. Are you relying on this for the check to work? If so please add a comment here.

File: client/script/client.js [hunk 3]
```diff
@@ -17,11 +19,41 @@ var client = (function () {
...
+    var $parent = element.parent();
+
+    if (element.val().trim().toLowerCase().match(pattern)) {
+      $parent.removeClass('has-error').addClass('has-success has-feedback');
+      $parent.find('.success-icon').removeClass('hidden').addClass('show');
+      $parent.find('.fail-icon').removeClass('show').addClass('hidden');
+      $parent.find('.fail-help').removeClass('show').addClass('hidden');
+      $parent.find('.fail-custom').removeClass('show').addClass('hidden');
+      if (checkServerFlag)
+        verifyKeysAndFetchDescription(function(_) { });
```
Comment (frederickjansen) [inline client/script/client.js]: Don't add empty callback here, instead check if the callback exists in the method itself

File: client/script/client.js [hunk 4]
```diff
@@ -17,11 +19,41 @@ var client = (function () {
...
+    }
+  };
+
+  function verifyKeysAndFetchDescription(callback) {
     var session = $("#session").val().trim().toLowerCase();
     var participationCode = $("#participation-code").val().trim().toLowerCase();
 
-    if (session === "" || participationCode === "") return;
+    if (session == "" || participationCode == "") {
+      callback(false);
```
Comment (frederickjansen) [inline client/script/client.js:55]: Change to ```callback && callback(false)```

File: client/script/client.js [hunk 5]
```diff
@@ -63,67 +97,85 @@ var client = (function () {
...
     }
 
-    // Verify confirmation check box was checked
-    var verifyChecked = $('#verify').is(':checked');
-    if (!verifyChecked) {
-      errors = errors.concat(UNCHECKED_ERR);
-    }
+    // Validate the remaining components after session and
+    // and participation code are validated with the server.
+    var validate_remaining_components = function(result) {
```
Comment (frederickjansen) [inline client/script/client.js]: camelCase

File: client/script/client.js [hunk 6]
```diff
@@ -1,6 +1,8 @@
 var client = (function () {
-  var SESSION_KEY_ERROR = 'Invalid session number: must be 26-character combination of letters and numbers';
-  var PARTICIPATION_CODE_ERROR = 'Invalid participation code: must be 26-character combination of letters and numbers';
+  var SESSION_KEY_ERROR = 'Invalid session number';
+  var PARTICIPATION_CODE_ERROR = 'Invalid participation code';
+
+  var SESSION_PARTICIPATION_CODE_SERVER_ERROR = 'Session number and participation code do not match';
```
Comment (frederickjansen) [inline client/script/client.js:5]: This isn't used anywhere

File: client/script/client.js [hunk 7]
```diff
@@ -43,6 +75,7 @@ var client = (function () {
       $parent.find('.fail-icon').removeClass('show').addClass('hidden');
       $parent.find('.fail-help').removeClass('show').addClass('hidden');
       $parent.find('.fail-custom').removeClass('show').addClass('hidden');
+      callback(true);
```
Comment (frederickjansen) [inline client/script/client.js:79]: Same thing here and on line 90

**Review by frederickjansen — CHANGES_REQUESTED** [review]
I'm going to merge as-is, comments can be addressed after the demo.

File: client/script/client.js (hunk 2 above)
Comment (KinanBab) [reply client/script/client.js:54]: did not do this intentionally.

**Pull request merged**